runtime/src/bank/builtins/core_bpf_migration/target_core_bpf.rs

```diff
@@ -0,0 +1,295 @@
+use {
+    super::error::CoreBpfMigrationError,
+    crate::bank::Bank,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        bpf_loader_upgradeable::{self, get_program_data_address, UpgradeableLoaderState},
+        pubkey::Pubkey,
+    },
+};
+
+/// The account details of a Core BPF program slated to be upgraded.
+#[derive(Debug)]
+pub(crate) struct TargetCoreBpf {
+    pub program_address: Pubkey,
+    pub program_account: AccountSharedData,
+    pub program_data_address: Pubkey,
+    pub program_data_account: AccountSharedData,
+}
+
+impl TargetCoreBpf {
+    /// Collects the details of a Core BPF program and verifies it is properly
+    /// configured.
+    /// The program account should exist with a pointer to its data account.
+    /// The program data account should exist with the correct state
+    /// (a ProgramData header and the program ELF).
+    pub(crate) fn new_checked(
+        bank: &Bank,
+        program_address: &Pubkey,
+    ) -> Result<Self, CoreBpfMigrationError> {
+        let program_data_address = get_program_data_address(program_address);
+
+        // The program account should exist.
+        let program_account = bank
+            .get_account_with_fixed_root(program_address)
+            .ok_or(CoreBpfMigrationError::AccountNotFound(*program_address))?;
+
+        // The program account should be owned by the upgradeable loader.
+        if program_account.owner() != &bpf_loader_upgradeable::id() {
+            return Err(CoreBpfMigrationError::IncorrectOwner(*program_address));
+        }
+
+        // The program account should have a pointer to its data account.
+        match program_account.deserialize_data::<UpgradeableLoaderState>()? {
+            UpgradeableLoaderState::Program {
+                programdata_address,
+            } if programdata_address == program_data_address => (),
+            _ => {
+                return Err(CoreBpfMigrationError::InvalidProgramAccount(
+                    *program_address,
+                ))
+            }
+        }
+
+        // The program data account should exist.
+        let program_data_account = bank
+            .get_account_with_fixed_root(&program_data_address)
+            .ok_or(CoreBpfMigrationError::ProgramHasNoDataAccount(
+                *program_address,
+            ))?;
+
+        // The program data account should be owned by the upgradeable loader.
+        if program_data_account.owner() != &bpf_loader_upgradeable::id() {
+            return Err(CoreBpfMigrationError::IncorrectOwner(program_data_address));
+        }
+
+        // The program data account should have the correct state.
+        let programdata_metadata_size = UpgradeableLoaderState::size_of_programdata_metadata();
+        if program_data_account.data().len() >= programdata_metadata_size {
+            if let UpgradeableLoaderState::ProgramData { .. } =
```

Comment:
I'm thinking we should also check that `upgrade_authority_address` is None. I don't think it makes sense to support this path for programs that are also independently upgradeable. Also, we probably want to prevent magically setting a new upgrade authority via a feature-gated upgrade.

Reply:
Agreed on the magic replacement bit, but as for your first point, couldn't there still be a situation where an upgradeable Core BPF program needs to be upgraded in sync with a runtime feature activation? Couldn't there be a feature activation that could break a program if it's not upgraded?

Reply:
> couldn't there still be a situation where an upgradeable Core BPF program needs to be upgraded in sync with a runtime feature activation? Couldn't there be a feature activation that could break a program if it's not upgraded?

Presumably we would use the same feature gate to sync the program upgrade with the runtime behavior change